File: javascript/21_crud/lista_crud.js
const lista = (function () {
	var cabeca = {};
	var topo = 0;

	function obter(posicao) {
		var elemento = cabeca;

		for (var i = 0; i < posicao; i++) {
			elemento = elemento.proximo;
		}

		return elemento;
	}

	function inserir(nome, preco, posicao) {
		if (posicao >= 0 && posicao <= topo) {
			var novo = { nome, preco, proximo: null };

			if (posicao == 0) {
				novo.proximo = cabeca;
				cabeca = novo;
			} else {
				var antecessor = obter(posicao - 1);
				var sucessor = antecessor.proximo;

				antecessor.proximo = novo;
				novo.proximo = sucessor;
			}

			topo++;

			return true;
		}

		return false;
	}

	function remover(posicao) {
		if (posicao == 0) {
			cabeca = cabeca.proximo;
		} else {
			var antecessor = obter(posicao - 1);
			var corrente = antecessor.proximo;
			var sucessor = corrente.proximo;

			antecessor.proximo = sucessor;
		}

		topo--;
	}

	function inserirEmOrdem(nome, preco) {
		var corrente = cabeca;

		if (topo == 0) {
			inserir(nome, preco, 0);
			return;
		}

		var i = 0;

		for (; i < topo; i++) {
			if (nome < corrente.nome) {
				break;
			}

			corrente = corrente.proximo;
		}

		inserir(nome, preco, i);
	}

	function imprimir() {
		if (topo >= 0) {
			var i = 1;
			var corrente = cabeca;

			console.log("-------itens-------");

			do {
				console.log(i + ". " + corrente.nome + ": " + corrente.preco);
				corrente = corrente.proximo;
				i++;
			} while (corrente != null);
		}
	}

	return {
		remover: remover,
		obter: obter,
		imprimir: imprimir,
		inserirEmOrdem: inserirEmOrdem
	}
}()); 

File: python/_modulos/lista_crud.py
import json # para salvar dados

class Lista:
    def __init__(self):
        self.cabeca = None
        self.topo = 0

    def obter(self, posicao):
        elemento = self.cabeca

        for i in range(posicao):
            elemento = elemento["proximo"]

        return elemento
    
    def inserir(self, novo, posicao):
        if (posicao >= 0) and (posicao <= self.topo):
            if posicao == 0:
                novo["proximo"] = self.cabeca
                self.cabeca = novo
            else:
                antecessor = self.obter(posicao - 1)
                sucessor = antecessor["proximo"]

                antecessor["proximo"] = novo
                novo["proximo"] = sucessor

            self.topo += 1
            return True
        
        return False
    
    def remover(self, posicao):
        if posicao == 0:
            self.cabeca = self.cabeca["proximo"]
        else:
            antecessor = self.obter(posicao - 1)
            corrente = antecessor["proximo"]
            sucessor = corrente["proximo"]
            
            antecessor["proximo"] = sucessor        
        
        self.topo -= 1

    def inserirEmOrdem(self, novo):
        corrente = self.cabeca

        if self.topo == 0:
            self.inserir(novo, 0)
            return
        
        i = 0
        while i < self.topo:
            if novo["nome"] < corrente["nome"]:
                break

            corrente = corrente["proximo"]
            i += 1

        self.inserir(novo, i)

    def imprimir(self):
        if self.topo > 0:
            i = 1
            corrente = self.cabeca

            print("-------itens-------")

            while corrente != None:
                print(f"{i}. {corrente['nome']}: {corrente['preco']}")
                corrente = corrente["proximo"]
                i += 1

    def salvar(self):
        arquivo = open("produtos.json", "w")
        dados = {"tamanho": self.topo, "produtos": self.cabeca}
        json.dump(dados, arquivo)
        arquivo.close()

    def carregar(self):
        arquivo = open("produtos.json", "r")
        dados = json.load(arquivo)
        self.topo = dados["tamanho"]
        self.cabeca = dados["produtos"]
        arquivo.close()